lib/id3c/cli/redcap.py

```diff
@@ -337,3 +336,40 @@ def is_complete(instrument: str, data: dict) -> bool:
         InstrumentStatus.Complete.value,
         str(InstrumentStatus.Complete.value)
     }
+
+
+def url_endpoints(url) -> Tuple[str, str]:
+    """
+    Returns a tuple of ``(api_url, base_url)`` based on the given REDCap
+    instance *url*.
+
+    *url* may be either the API endpoint or the base path of the REDCap
+    instance.
+
+    >>> url_endpoints("https://redcap.iths.org/")
+    ('https://redcap.iths.org/api/', 'https://redcap.iths.org/')
+
+    >>> url_endpoints("https://redcap.iths.org/api/")
+    ('https://redcap.iths.org/api/', 'https://redcap.iths.org/')
+
+    >>> url_endpoints("https://example.org/redcap/")
+    ('https://example.org/redcap/api/', 'https://example.org/redcap/')
+
+    >>> url_endpoints(url_endpoints("https://redcap.iths.org/api/")[1])
+    ('https://redcap.iths.org/api/', 'https://redcap.iths.org/')
+
+    Assumes that the API endpoint is always at ``…/api/`` under the REDCap
+    instance's base URL and vice versa.  This appears to be true for all the
+    instances inspected (redcap.iths.org, redcap.fhcrc.org,
+    redcapdemo.vanderbilt.edu).
+    """
+    url = Url(url)
+
+    if re.search(r'(?<=/)api/?$', url.path):
```

Comment:
Do you not need to escape the `/` characters in this regular expression?

Reply:
I only ask because I entered this into an online regex evaluator, and it suggested that the regex was invalid without escaping them.

Reply:
In other languages, `/` are syntax for delimiting a regex literal, the same way `'` are syntax for delimiting a string literal. In those languages, you do have to escape `/` in a pattern, because the `/` would be treated as a delimiter otherwise. For example, in JS, you'd write this code something like:

```js
if (url.path.match(/(?<=\/)api\/?$/)) {
```

Some languages even let you use different delimiters for regex literals to avoid having to verbosely escape your `/` if you have a lot of them in your pattern.

Even in those languages, you can often also provide your regex patterns as a string, but then they're not regex literals, e.g. in JS:

```js
if (url.path.match('(?<=/)api/?$')) {
```

However, Python doesn't use `/` to delimit regex literals, because it doesn't support regex literals to begin with. You have to provide your regex as a string, as I do here with `r'…'`.

Reply:
Thanks, Tom, for laying this out. I appreciate it.